src/System.Management.Automation/engine/parser/TypeInferenceVisitor.cs

```diff
@@ -0,0 +1,1572 @@
+
+using CimClass = Microsoft.Management.Infrastructure.CimClass;
+using CimInstance = Microsoft.Management.Infrastructure.CimInstance;
+
+namespace System.Management.Automation
+{
+    /// <summary>
+    /// Enum describing permissions to use runtime evaluation during type inference
+    /// </summary>
+    [Flags]
+    public enum TypeInferenceRuntimePermissions {
+        /// <summary>
+        /// No runtime use is allowed
+        /// </summary>
+        None = 0,
+        /// <summary>
+        /// Use of SafeExprEvaluator visitor is allowed
+        /// </summary>
+        AllowSafeEval = 1,
+    }
+
+    /// <summary>
+    /// static class containing methods to work with type inference of abstract syntax trees
+    /// </summary>
+    internal static class AstTypeInference
+    {
+        /// <summary>
+        /// Infers the type that the result of executing a statement would have without using runtime safe eval
+        /// </summary>
+        /// <param name="ast">the ast to infer the type from</param>
+        /// <returns></returns>
+        public static IList<PSTypeName> InferTypeOf(Ast ast)
+        {
+            return InferTypeOf(ast, TypeInferenceRuntimePermissions.None);
+        }
+
+        /// <summary>
+        /// Infers the type that the result of executing a statement would have
+        /// </summary>
+        /// <param name="ast">the ast to infer the type from</param>
+        /// <param name="evalPermissions">The runtime usage permissions allowed during type inference</param>
+        /// <returns></returns>
+        public static IList<PSTypeName> InferTypeOf(Ast ast, TypeInferenceRuntimePermissions evalPermissions)
+        {
+            return InferTypeOf(ast, PowerShell.Create(RunspaceMode.CurrentRunspace), evalPermissions);
+        }
+
+        /// <summary>
+        /// Infers the type that the result of executing a statement would have without using runtime safe eval
+        /// </summary>
+        /// <param name="ast">the ast to infer the type from</param>
+        /// <param name="powerShell">the instance of powershell to user for expression evalutaion needed for type inference</param>
+        /// <returns></returns>
+        public static IList<PSTypeName> InferTypeOf(Ast ast, PowerShell powerShell)
+        {
+            return InferTypeOf(ast, TypeInferenceRuntimePermissions.None);
+        }
+
+        /// <summary>
+        /// Infers the type that the result of executing a statement would have
+        /// </summary>
+        /// <param name="ast">the ast to infer the type from</param>
+        /// <param name="powerShell">the instance of powershell to user for expression evalutaion needed for type inference</param>
+        /// <param name="evalPersmissions">The runtime usage permissions allowed during type inference</param>
+        /// <returns></returns>
+        public static IList<PSTypeName> InferTypeOf(Ast ast, PowerShell powerShell,TypeInferenceRuntimePermissions evalPersmissions)
```

Comment:
Better add comments and in the constructors to call it out.
Do you expect the passed in powershell instance to associate with a remote runspace or a runspace pool? For the usage of it in our tab completion code, it will always be created from `powershell.Create(CurrentRunspace)`, but I'm having trouble to figure out the scenarios it will be used outside tab completion code.